File: C/3/3.c
#include <stdio.h>

/* Control string size */
#define LIMIT 999

/**
 * A function that expands shorthand notations like a-z in the string s1 into
 * the equivalent complete list abc...xyz in s2.
 *
 * @param char[] s1 The string with the shorthand notation
 * @param char[] s2 The expanded string
 *
 * @return void
 */
void expand(char s1[], char s2[])
{
    int i = 0, j = 0, c = 0, first = -1, last = -1;

    /* Dealing with leading dash. */
    if (s1[0] == '-') {
        i++;
    }

    while(s1[i] != '\0') {
        /* Find first character of sequence. */
        first = s1[i++];
        /**
         * If we are on a dash and the char after the dash is greater than the
         * character before, we have a range.
         */
        last = (s1[i] == '-' && s1[i+1] > s1[i-1])? s1[++i] : -1;

        /* If a range could be defined. */
        if (first != -1 && last != -1) {
            for (c = first; c <= last; j++, c++) {
                s2[j] = c;
            }
            s2[j] = '\0';
            first = last = -1;
        }
    }

    return;
}

/**
 * A program to test string expansion.
 */
main()
{
    char input1[] = "a-z";
    char output1[LIMIT];

    char input2[] = "a-b-c";
    char output2[LIMIT];

    char input3[] = "a-z0-9";
    char output3[LIMIT];

    char input4[] = "-a-z";
    char output4[LIMIT];

    expand(input1, output1);
    printf("expand('%s') => '%s'\n", input1, output1);

    expand(input2, output2);
    printf("expand('%s') => '%s'\n", input2, output2);

    expand(input3, output3);
    printf("expand('%s') => '%s'\n", input3, output3);

    expand(input4, output4);
    printf("expand('%s') => '%s'\n", input4, output4);
}

File: rustlang/3/3.rs
/// A function that expands shorthand notations like a-z in the string s1 into
/// the equivalent complete list abc...xyz in s2.
///
/// @param &str s1 The string with the shorthand notation
///
/// @return String The expanded string
fn expand(s1 : &str) -> String
{
    let mut result = String::new();

    // We use a peekable iterator, so we can check characters ahead
    let mut iter = s1.chars().peekable();

    // Dealing with leading dash.
    if iter.peek() == Some(&&'-'){
        iter.next();
    }

    loop {
        // Find first character of sequence.
        let first = iter.next();
        let last = if iter.peek() == Some(&&'-') {
                        iter.next();
                        // Just peek, without removing, because it might be the start of a new sequence later on
                        iter.peek()
                    } else {
                        // We have to return a reference here because that's what peek returns too
                        Some(&'-')
                    };

        if first == None {
            // No more characters to be read.
            break;
        } else if last != Some(&'-') {
            // iter.next() and iter.peek() both return options, so we have to unwrap them
            let begin : u8 = first.unwrap() as u8;
            // iter.peek() returns a reference, so we need to dereference to unwrap (E0606)
            let end : u8 = *last.unwrap() as u8;

            // Push characters to resultant array
            for _c in begin..end+1 {
                result.push(_c as char);
            }
        }
    }

    result
}

/// A program to test string expansion.
fn main()
{
    let input1 = String::from("a-z");

    let input2 = String::from("a-b-c");

    let input3 = String::from("a-z0-9");

    let input4 = String::from("-a-z");

    print!("expand('{}') => '{}'\n", input1, expand(&input1));

    print!("expand('{}') => '{}'\n", input2, expand(&input2));

    print!("expand('{}') => '{}'\n", input3, expand(&input3));

    print!("expand('{}') => '{}'\n", input4, expand(&input4));
}
